language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/providers/AddOnFailCodeAction.java

```diff
@@ -0,0 +1,116 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.ballerinalang.langserver.codeaction.providers;
+
+import io.ballerina.compiler.syntax.tree.CheckExpressionNode;
+import io.ballerina.compiler.syntax.tree.DoStatementNode;
+import io.ballerina.compiler.syntax.tree.Node;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.VariableDeclarationNode;
+import io.ballerina.tools.diagnostics.Diagnostic;
+import io.ballerina.tools.text.LinePosition;
+import org.ballerinalang.annotation.JavaSPIService;
+import org.ballerinalang.langserver.common.constants.CommandConstants;
+import org.ballerinalang.langserver.common.utils.CommonUtil;
+import org.ballerinalang.langserver.commons.CodeActionContext;
+import org.ballerinalang.langserver.commons.codeaction.spi.DiagBasedPositionDetails;
+import org.ballerinalang.util.diagnostic.DiagnosticErrorCode;
+import org.eclipse.lsp4j.CodeAction;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.TextEdit;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * On fail suggestion Code Action for check expressions.
+ *
+ * @since 2.0.0
+ */
+@JavaSPIService("org.ballerinalang.langserver.commons.codeaction.spi.LSCodeActionProvider")
+public class AddOnFailCodeAction extends AbstractCodeActionProvider {
+
+    public static final String NAME = "Add on-fail clause";
+
+    @Override
+    public List<CodeAction> getDiagBasedCodeActions(Diagnostic diagnostic,
+                                                    DiagBasedPositionDetails positionDetails,
+                                                    CodeActionContext context) {
+
+        if (!(DiagnosticErrorCode.CHECKED_EXPR_NO_MATCHING_ERROR_RETURN_IN_ENCL_INVOKABLE.diagnosticId()
+                .equals(diagnostic.diagnosticInfo().code()))) {
+            return Collections.emptyList();
+        }
+
+        CheckExpressionNode checkExpressionNode = (CheckExpressionNode) positionDetails.matchedNode();
+        
+        if (isDoStmtNodePresent(positionDetails.matchedNode())) {
+            DoStatementNode doStatementNode = getDoStatementNode(checkExpressionNode);
+            LinePosition doStmtLinePosition = doStatementNode.lineRange().endLine();
+            Position onFailPosLine = new Position(doStmtLinePosition.line(), doStmtLinePosition.offset());
+            
+            String spaces = " ".repeat(doStmtLinePosition.offset() - 1);
+            String editText = " on fail var e {" + CommonUtil.LINE_SEPARATOR + spaces + "\t"
+                + CommonUtil.LINE_SEPARATOR + spaces + "}";
+            List<TextEdit> edits = Arrays.asList(new TextEdit(new Range(onFailPosLine, onFailPosLine), editText));
```

Comment:
Use `Collections.emptyList`

Reply:
Shouldn't `edits` be a `Collections.singletonList` as well since it has an element?

Reply:
Yes, should be `singletonList.`.

Reply:
Ayye I was thinking refactoring the logic as following
```
List<TextEdit> edits = new ArrayList<>();
if (logic) {
    edits.add(new TextEdit());
} else {
    edits.add(new TextEdit());
}
return Collections.singletonList();
```
where I have written this as follows (after using singletonList)
```
if (logic) {
    List<TextEdit> edits = Collections.singletonList(new TextEdit());
    edits.add(new TextEdit());
    return Collections.singletonList(createQuickFix(edits));
} else {
    List<TextEdit> edits = Collections.singletonList(new TextEdit());
    edits.add(new TextEdit());
    return Collections.singletonList(createQuickFix(edits));
}
```
WDYT?

Reply:
If it's only the edits are changing and code action title is same in both cases, +1

Reply:
They have two different code action titles. which I have initialized prior to the if statement 
```
List<TextEdit> edits = new ArrayList<>();
String commandTitle;
if (logic) {
    edits.add(new TextEdit());
    commanTitle = "command1";
} else {
    edits.add(new TextEdit());
    commanTitle = "command2";
}
return Collections.singletonList();
```